tests/kernel/alert/alert_api/src/test_alert_contexts.c

```diff
@@ -224,6 +238,144 @@ void test_isr_kinit_alert(void)
+ *  1. The current task times out while waiting for the event.
+ *  2. There is already an event waiting (signalled from a task).
+ *  3. The current task must wait on the event until it is signalled
+ *     from either another task or an ISR.
+ */
+void test_thread_alert_timeout(void)
+{
+	/**TESTPOINT: alert handler ignore*/
+	struct k_alert alert;
+	int ret, i;
+
+	/**TESTPOINT: init via k_alert_init*/
+	k_alert_init(&alert, K_ALERT_DEFAULT, PENDING_MAX);
+
+	palert = &alert;
+
+	ret = k_alert_recv(&alert, TIMEOUT);
+
+	zassert_equal(ret, -EAGAIN, NULL);
+
+	k_alert_send(&alert);
+
+	ret = k_alert_recv(&alert, TIMEOUT);
+
+	zassert_equal(ret, 0, NULL);
+
+	k_sem_give(&sync_sema);
+
+	for (i = 0; i < 2; i++) {
+		ret = k_alert_recv(&alert, TIMEOUT);
+
+		zassert_equal(ret, 0, NULL);
+	}
+}
+
+/**
+ * This test checks alert_recv(K_FOREVER) against
+ * the following cases:
+ *  1. There is already an event waiting (signalled from a task and ISR).
+ *  2. The current task must wait on the event until it is signalled
+ *     from either another task or an ISR.
+ */
+void test_thread_alert_wait(void)
+{
+	/**TESTPOINT: alert handler ignore*/
+	struct k_alert alert;
+	int ret, i;
+
+	/**TESTPOINT: init via k_alert_init*/
+	k_alert_init(&alert, K_ALERT_DEFAULT, PENDING_MAX);
+
+	palert = &alert;
+
+	k_alert_send(&alert);
+
+	ret = k_alert_recv(&alert, K_FOREVER);
+
+	zassert_equal(ret, 0, NULL);
+
+	irq_offload(sync_entry, NULL);
+
+	ret = k_alert_recv(&alert, K_FOREVER);
+
+	zassert_equal(ret, 0, NULL);
+
+	k_sem_give(&sync_sema);
+
+	for (i = 0; i < 2; i++) {
+		ret = k_alert_recv(&alert, K_FOREVER);
+
+		zassert_equal(ret, 0, NULL);
+	}
+}
+
+int eventHandler(struct k_alert *alt)
+{
+	return handler_val;
+}
+
+/**
+ * This test checks that the event handler is set up properly when
+ * alert_event_handler_set() is called.  It shows that event handlers
+ * are tied to the specified event and that the return value from the
+ * handler affects whether the event wakes a task waiting upon that
+ * event.
+ */
+void test_thread_alert_handler(void)
+{
+	/**TESTPOINT: alert handler ignore*/
+	struct k_alert alert;
+	int ret;
+
+	/**TESTPOINT: init via k_alert_init*/
+	k_alert_init(&alert, eventHandler, PENDING_MAX);
+
+	palert = &alert;
+
+	k_sem_give(&sync_sema);
+
+	ret = k_alert_recv(&alert, TIMEOUT);
+
+	zassert_equal(ret, -EAGAIN, NULL);
+
+	k_sem_give(&sync_sema);
+
+	ret = k_alert_recv(&alert, TIMEOUT);
+
+	zassert_equal(ret, 0, NULL);
+}
+
+
+/**
+ * Alternate task to signal various events to a waiting task
+ */
+void AlternateTask(void *p1, void *p2, void *p3)
+{
+	k_sem_init(&sync_sema, 0, 1);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	k_alert_send(palert);
+	irq_offload(sync_entry, NULL);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	k_alert_send(palert);
+	irq_offload(sync_entry, NULL);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	handler_val = 0;
+	k_alert_send(palert);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	handler_val = 1;
+	k_alert_send(palert);
```

Comment:
All the threads/tests would have exited by the time this thread reaches this point. Who is waiting for this alert?

Reply:
no. When a handler is installed for a alert object , alert is sent only when handler return other than zero. Hence when handler_val = 0  k_alert_send has no effect. Hence thread will still be waiting.